Increase information in logs if a wrong secret is used

nicksinger opened the pull request:
Previous log output:
`[Mon Jun 19 12:55:35 2017] [18947:error] hmac check failed`

Log output with this PR:
`[Mon Jun 19 13:24:35 2017] [21137:error] Wrong api secret for key 'AABBCCDDEEFF0011' (belongs to user 'Demo') used.`

The error message only gets raised if the used secret for an existing key is wrong (hence the previous "hmac check failed"). If the log level is on debug, this information is also recorded in previous error messages. However if openQA runs on big production instances it is maybe not possible all the time to correlate previous messages correctly or it is not feasible to run with "debug". Therefore the increase of verbosity inside this specific message.

okurz commented:
LGTM, needs tidy (at least)

okurz commented:
on second thought, I think it should not be a warning but only an info. It is not an error within openQA that the user is using the wrong key (or a brute force attack is happening).

foursixnine commented on lib/OpenQA/WebAPI/Controller/API/V1.pm:
```diff
@@ -53,7 +53,7 @@ sub auth {
             }
             else {
-                $self->app->log->error("hmac check failed");
+                $self->app->log->error(sprintf "Wrong api secret for key '%s' (belongs to user '%s') used.", $api_key->key, $api_key->user->username);
```
Please also log the ip address, and keep in mind that we might be behind a proxy :)

coolo commented:
'hmac check failed' is actually most of the time because of broken NTP setup. better add that info to the log so that curious admins know what to fix. The rest is part of the apache access_log and there is no reason to duplicate that.

nicksinger commented:
So I changed the log level of this message to "warn". I think @okurz is right here and we should use "error" only for real backend failures. I also added the IP of the client which issues the wrong api requests into the logs so one could make use of fail2ban or similar tools.

@coolo I agree on having a hint to a (maybe) wrong NTP setup. Where? Pitfalls? Another log line on "warn" level (maybe only info)? Adding the additional information inside this log is IMHO still quiet useful since only special configured apache instances log the parameters of post requests too (which is required to see the used API key).

nicksinger commented:
43bdd47 contains the latest suggestion worked out by @okurz and me together. We tried to specify a reasonable log level and message depending on which path the code goes for verification of the secret.
It should now be easier for an administrator to decide if and what action he has to take if such log messages appear.
As always; feel free to suggest improvements in the handling and wording :)

coolo commented:
fixing the tests and removing the warnings during 03-auth.t would be a start

kraih commented on lib/OpenQA/WebAPI/Controller/API/V1.pm:
```diff
@@ -49,16 +49,26 @@ sub auth {
+                  $username, $request_ip, $api_key->key, $api_key->secret;
                 if (!_is_timestamp_valid($build_tx_timestamp, $timestamp)) {
                     $reason = "timestamp mismatch";
+                    $self->app->log->warn($log_msg . ", $reason");
```
Style nitpick: Since there are so many calls to `$self->app->log` in this method, it would make sense to assign the object to a variable somewhere at the top (`my $log = $self->app->log;`). Method calls in Perl are rather expensive, maybe not enough to make much of a difference here, but it's good practice.

nicksinger commented:
2c8090d took care of the failing tests @coolo mentioned. f2298e3 is the refactor @kraih mentioned.
The other failing travis jobs are caused by something different as far as i understand them.

okurz commented:
LGTM

nicksinger commented:
If nobody has something to add here, feel free to merge - from my site this PR is done,

coolo commented:
please rebase

coolo commented:
test fails

nicksinger commented:
Somehow my squash required another tidy. Travis pending

coolo commented:
I restarted the travis job - let's see
